File: 0202-happy-number/0202-happy-number.java
class Solution {
    private int check(int n){
        int sum = 0;
        while(n > 0){
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }
    public boolean isHappy(int n) {
        int slow = n;
        int fast = check(n);

        while(fast != 1 && slow != fast){
            slow = check(slow);
            fast = check(check(fast));
        }
        return fast == 1;
    }
}

File: 0202-happy-number/0202-happy-number.py
class Solution(object):
    def isHappy(self, n):
        def next_number(num):
            total = 0
            while num:
                num, digit = divmod(num, 10)
                total += digit * digit
            return total

        slow, fast = n, next_number(n)
        while fast != 1 and slow != fast:
            slow = next_number(slow)
            fast = next_number(next_number(fast))
        return fast == 1
